### PR TITLE
Template Strings

### LINKED ISSUE #1074: ES6 Feature: Template Strings: tracking Issue
### Template Literal

```
TemplateLiteral[Yield] :
  NoSubstitutionTemplate
  TemplateHead Expression[In, ?Yield] TemplateSpans[?Yield]
TemplateSpans[Yield] :
  TemplateTail
  TemplateMiddleList[?Yield] TemplateTail
TemplateMiddleList[Yield] :
  TemplateMiddle Expression[In, ?Yield]
  TemplateMiddleList[?Yield] TemplateMiddle Expression[In, ?Yield]
```

### Template Literal Lexical Components

```
Template ::
  NoSubstitutionTemplate
  TemplateHead
NoSubstitutionTemplate ::
  ` TemplateCharactersopt `
TemplateHead ::
  ` TemplateCharactersopt ${
TemplateSubstitutionTail ::
  TemplateMiddle
  TemplateTail
TemplateMiddle ::
  } TemplateCharactersopt ${
TemplateTail ::
  } TemplateCharactersopt `
TemplateCharacters ::
  TemplateCharacter TemplateCharactersopt
TemplateCharacter ::
  $ [lookahead ≠ { ]
  \ EscapeSequence
  LineContinuation
  LineTerminatorSequence
  SourceCharacter but not one of ` or \ or $ or LineTerminator
```

## Spec:

https://people.mozilla.org/~jorendorff/es6-draft.html#sec-template-literals
https://people.mozilla.org/~jorendorff/es6-draft.html#sec-template-literal-lexical-components

## Remaining Tasks:

- [x] Review and Test Token Generation
- [x] Review Error Messages 
- [x] Fix #1002
- [x] estree/estree#25

### PR DESCRIPTION
**Pull request by mikesherov**
Closes #1074

### CODE REVIEW

File: esprima.js [hunk 1]
```diff
@@ -234,6 +240,34 @@
         return '01234567'.indexOf(ch) >= 0;
     }
 
+    function octalToDecimal(ch) {
+        var code = '01234567'.indexOf(ch), octal = false;
+
+        // \0 is not octal escape sequence
+        if (code !== 0) {
+            octal = true;
+        }
```
Comment (ariya) [inline esprima.js]: Simpler:

```js
        var octal = (ch !== '0'), code = '01234567'.indexOf(ch);
```

File: esprima.js [hunk 2]
```diff
@@ -234,6 +240,34 @@
...
+
+        /* istanbul ignore else */
```
Comment (ariya) [inline esprima.js]: I don't think this is necessary. The original code didn't skip the else branch.

File: esprima.js [hunk 3]
```diff
@@ -1402,6 +1542,11 @@
             return scanStringLiteral();
         }
 
+        // Template literals start with ` (U+0060) for template head
+        // or } (U+007D) for template middle or template tail.
+        if (ch === 0x60 || (ch === 0x7D && state.curlyStack[state.curlyStack.length - 1] === '${')) {
+            return scanTemplate();
+        }
```
Comment (ariya) [inline esprima.js]: I think template literals are not so common yet. Thus, better to have this check as the last one (before `return scanPunctuator()`) for minor performance gain.

**Comment by ariya** [comment]
Other than the minor tweaks, LGTM :+1:

File: esprima.js (hunk 3 above)
Comment (mikesherov) [inline esprima.js]: addressed.

**Comment by mikesherov** [comment]
@ikarienator can you re-review? Good catch on the octal literals. I missed that part from the annex.

File: esprima.js [hunk 4]
```diff
@@ -1027,26 +1060,10 @@
 
                     default:
                         if (isOctalDigit(ch)) {
```
Comment (ikarienator) [inline esprima.js]: Please remember to change this as we've discussed in harmony branch.

File: test/fixtures/ES6/template-literals/literal-escape-sequences.source.js [hunk 5]
```diff
@@ -0,0 +1 @@
+var source = '`\\u{000042}\\u0042\\x42\\u0\\A\\0`';
```
Comment (ikarienator) [inline test/fixtures/ES6/template-literals/literal-escape-sequences.source.js]: Please fix the truncated `\u`.

**Comment by ikarienator** [comment]
Minor issue otherwise LGTM :+1: The tests cases are very impressive.

**Comment by ikarienator** [comment]
@mikesherov Hi, do you have time to work on this?

**Comment by mikesherov** [comment]
@ikarienator yes, it'll be done before the next meeting.

File: esprima.js (hunk 4 above)
Comment (ariya) [inline esprima.js]: @ikarienator Could you point me to the discussion?

**Pull request closed without merging**